data/src/isupport.rs

```diff
@@ -0,0 +1,614 @@
+    FNC,
+    HOSTLEN(u16),
+    INVEX(char),
+    KEYLEN(u16),
+    KICKLEN(u16),
+    KNOCK,
+    LINELEN(u16),
+    MAP,
+    MAXBANS(u16),
+    MAXCHANNELS(u16),
+    MAXLIST(Vec<ModesLimit>),
+    MAXPARA(u16),
+    MAXTARGETS(Option<u16>),
+    METADATA(Option<u16>),
+    MODES(Option<u16>),
+    MONITOR(Option<u16>),
+    MSGREFTYPES(Vec<MessageReferenceType>),
+    NAMESX,
+    NETWORK(String),
+    NICKLEN(u16),
+    OVERRIDE,
+    PREFIX(Vec<PrefixMap>),
+    SAFELIST,
+    SECURELIST,
+    SILENCE(Option<u16>),
+    STATUSMSG(String),
+    TARGMAX(Vec<CommandTargetLimit>),
+    TOPICLEN(u16),
+    UHNAMES,
+    USERIP,
+    USERLEN(u16),
+    UTF8ONLY,
+    VLIST(String),
+    WATCH(u16),
+    WHOX,
+    Negation(String),
+}
+
+impl TryFrom<String> for Parameter {
+    type Error = &'static str;
+
+    fn try_from(isupport: String) -> Result<Self, Self::Error> {
+        Self::try_from(isupport.as_str())
+    }
+}
+
+impl<'a> TryFrom<&'a str> for Parameter {
+    type Error = &'static str;
+
+    fn try_from(isupport: &'a str) -> Result<Self, Self::Error> {
+        if isupport.is_empty() {
+            return Err("empty ISUPPORT parameter not allowed");
+        }
+
+        match isupport.chars().next() {
+            Some('-') => Ok(Parameter::Negation(isupport[1..].to_string())),
+            _ => {
+                if let Some((parameter, value)) = isupport.split_once('=') {
+                    match parameter {
+                        "ACCEPT" => Ok(Parameter::ACCEPT(parse_required_positive_integer(value)?)),
+                        "ACCOUNTEXTBAN" => {
+                            let account_based_extended_ban_masks =
+                                value.split(',').map(String::from).collect::<Vec<_>>();
+
+                            if !account_based_extended_ban_masks.is_empty() {
+                                Ok(Parameter::ACCOUNTEXTBAN(account_based_extended_ban_masks))
+                            } else {
+                                Err("no valid account-based extended ban masks")
+                            }
+                        }
+                        "AWAYLEN" => {
+                            Ok(Parameter::AWAYLEN(parse_optional_positive_integer(value)?))
+                        }
+                        "BOT" => Ok(Parameter::BOT(parse_required_letter(value, None)?)),
+                        "CALLERID" => Ok(Parameter::CALLERID(parse_required_letter(
+                            value,
+                            Some(default_caller_id_letter()),
+                        )?)),
+                        "CASEMAPPING" => match value {
+                            "ascii" => Ok(Parameter::CASEMAPPING(CaseMap::ASCII)),
+                            "rfc1459" => Ok(Parameter::CASEMAPPING(CaseMap::RFC1459)),
+                            "rfc1459-strict" => Ok(Parameter::CASEMAPPING(CaseMap::RFC1459_STRICT)),
+                            "rfc7613" => Ok(Parameter::CASEMAPPING(CaseMap::RFC7613)),
```

Comment:
are these always went lowercase or should we call `value.lowercase()` just in case?

Reply:
My reading is that they're always sent lowercase, but I don't see any reason we need to be strict about that.  Added `to_lowercase()`.